source/opt/inline_pass.cpp

```diff
@@ -0,0 +1,430 @@
+            SpvStorageClassFunction)
+      break;
+  }
+  return (type_itr != module_->types_values_end()) ? type_itr->result_id() : 0;
+}
+
+uint32_t InlinePass::AddPointerToType(uint32_t typeId) {
+  uint32_t resultId = TakeNextId();
+  std::vector<ir::Operand> in_operands;
+  in_operands.emplace_back(
+      spv_operand_type_t::SPV_OPERAND_TYPE_STORAGE_CLASS,
+      std::initializer_list<uint32_t>{uint32_t(SpvStorageClassFunction)});
+  in_operands.emplace_back(spv_operand_type_t::SPV_OPERAND_TYPE_ID,
+                           std::initializer_list<uint32_t>{uint32_t(typeId)});
+  std::unique_ptr<ir::Instruction> type_inst(
+      new ir::Instruction(SpvOpTypePointer, 0, resultId, in_operands));
+  module_->AddType(std::move(type_inst));
+  return resultId;
+}
+
+void InlinePass::AddBranch(uint32_t labelId,
+                           std::unique_ptr<ir::BasicBlock>* bp) {
+  std::vector<ir::Operand> branch_in_operands;
+  branch_in_operands.push_back(
+      ir::Operand(spv_operand_type_t::SPV_OPERAND_TYPE_ID,
+                  std::initializer_list<uint32_t>{labelId}));
+  std::unique_ptr<ir::Instruction> newBranch(
+      new ir::Instruction(SpvOpBranch, 0, 0, branch_in_operands));
+  (*bp)->AddInstruction(std::move(newBranch));
+}
+
+void InlinePass::AddStore(uint32_t ptrId, uint32_t valId,
+                          std::unique_ptr<ir::BasicBlock>* bp) {
+  std::vector<ir::Operand> store_in_operands;
+  store_in_operands.push_back(
+      ir::Operand(spv_operand_type_t::SPV_OPERAND_TYPE_ID,
+                  std::initializer_list<uint32_t>{ptrId}));
+  store_in_operands.push_back(
+      ir::Operand(spv_operand_type_t::SPV_OPERAND_TYPE_ID,
+                  std::initializer_list<uint32_t>{valId}));
+  std::unique_ptr<ir::Instruction> newStore(
+      new ir::Instruction(SpvOpStore, 0, 0, store_in_operands));
+  (*bp)->AddInstruction(std::move(newStore));
+}
+
+void InlinePass::AddLoad(uint32_t typeId, uint32_t resultId, uint32_t ptrId,
+                         std::unique_ptr<ir::BasicBlock>* bp) {
+  std::vector<ir::Operand> load_in_operands;
+  load_in_operands.push_back(
+      ir::Operand(spv_operand_type_t::SPV_OPERAND_TYPE_ID,
+                  std::initializer_list<uint32_t>{ptrId}));
+  std::unique_ptr<ir::Instruction> newLoad(
+      new ir::Instruction(SpvOpLoad, typeId, resultId, load_in_operands));
+  (*bp)->AddInstruction(std::move(newLoad));
+}
+
+// Generate callee code into newBlocks to be inlined for the function call at
+// call_ii. Also add new function variables into caller func
+
+void InlinePass::GenInlineCode(
+    std::vector<std::unique_ptr<ir::BasicBlock>>* newBlocks,
+    std::vector<std::unique_ptr<ir::Instruction>>* newVars,
+    ir::UptrVectorIterator<ir::Instruction> call_ii,
+    ir::UptrVectorIterator<ir::BasicBlock> call_bi) {
+  // Map from callee id to caller id
+  std::unordered_map<uint32_t, uint32_t> callee2caller;
+  // Pre-call OpSampledImage Insts
```

Comment:
I don't argue with your comment, although I must admit that I am having trouble understanding the "same block" rule for both OpImage and OpSampledImage. Maybe this isn't the place to have this discussion ultimately, but maybe I can bounce this off you first before taking it further.

It seems that any shader compiler can do the same analysis that I have to do to track down the instructions of interest when finally generating code. It seems to me it is sufficient that an OpSampledImage or OpImage must dominate its use. That plus prohibition against OpPhi and OpSelects (and OpStores?) having these as operands should be sufficient to guard against the kind of behavior that shader compiler writers are concerned about.

I object here partly because the addition of OpImage support to this algorithm will actually cause an escalation in the algorithm. Previously I could simply clone the OpSampledImage in the final block and be done, but now I have to scan the operands of the cloned instruction looking for OpImages. And I will have to scan OpImage operands looking for OpSampledImage. Basically, I will have to add an iterative/recursive component to the algorithm that will handle an alternating cascade of these instructions. This additional complication caused me to reconsider challenging the "same block" rule.

What do you think?